File: js/packages/teams-ai/src/prompts/PromptSectionBase.ts
/**
 * @module teams-ai
 */
/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */

import { Message, MessageContentParts } from "./Message";
import { PromptFunctions } from "./PromptFunctions";
import { PromptSection, RenderedPromptSection } from "./PromptSection";
import { TurnContext } from 'botbuilder';
import { Tokenizer } from "../tokenizers";
import { Memory } from "../MemoryFork";

/**
 * Abstract Base class for most prompt sections.
 * @remarks
 * This class provides a default implementation of `renderAsText()` so that derived classes only
 * need to implement `renderAsMessages()`.
 */
export abstract class PromptSectionBase implements PromptSection {
    public readonly required: boolean;
    public readonly tokens: number;
    public readonly separator: string;
    public readonly textPrefix: string;

    /**
     * Creates a new 'PromptSectionBase' instance.
     * @param tokens Optional. Sizing strategy for this section. Defaults to `auto`.
     * @param required Optional. Indicates if this section is required. Defaults to `true`.
     * @param separator Optional. Separator to use between sections when rendering as text. Defaults to `\n`.
     * @param textPrefix Optional. Prefix to use for text output. Defaults to `undefined`.
     */
    public constructor(tokens: number = -1, required: boolean = true, separator: string = '\n', textPrefix: string = '') {
        this.required = required;
        this.tokens = tokens;
        this.separator = separator;
        this.textPrefix = textPrefix;
    }

    /**
     * Renders the prompt section as a string of text.
     * @param context Context for the current turn of conversation.
     * @param memory Interface for accessing state variables.
     * @param functions Functions for rendering prompts.
     * @param tokenizer Tokenizer to use for encoding/decoding text.
     * @param maxTokens Maximum number of tokens allowed for the rendered prompt.
     * @returns The rendered prompt section.
     */
    public async renderAsText(context: TurnContext, memory: Memory, functions: PromptFunctions, tokenizer: Tokenizer, maxTokens: number): Promise<RenderedPromptSection<string>> {
        // Render as messages
        const asMessages = await this.renderAsMessages(context, memory, functions, tokenizer, maxTokens);
        if (asMessages.output.length === 0) {
            return { output: '', length: 0, tooLong: false };
        }

        // Convert to text
        let text = asMessages.output.map((message) => PromptSectionBase.getMessageText(message)).join(this.separator);

        // Calculate length
        const prefixLength = tokenizer.encode(this.textPrefix).length;
        const separatorLength = tokenizer.encode(this.separator).length;
        let length = prefixLength + asMessages.length + ((asMessages.output.length - 1) * separatorLength);

        // Truncate if fixed length
        text = this.textPrefix + text;
        if (this.tokens > 1.0 && length > this.tokens) {
            const encoded = tokenizer.encode(text);
            text = tokenizer.decode(encoded.slice(0, this.tokens));
            length = this.tokens;
        }

        return { output: text, length: length, tooLong: length > maxTokens };
    }

    /**
     * Renders the prompt section as a list of `Message` objects.
     * @remarks
     * MUST be implemented by derived classes.
     * @param context Context for the current turn of conversation.
     * @param memory Interface for accessing state variables.
     * @param functions Functions for rendering prompts.
     * @param tokenizer Tokenizer to use for encoding/decoding text.
     * @param maxTokens Maximum number of tokens allowed for the rendered prompt.
     * @returns The rendered prompt section.
     */
    public abstract renderAsMessages(context: TurnContext, memory: Memory, functions: PromptFunctions, tokenizer: Tokenizer, maxTokens: number): Promise<RenderedPromptSection<Message<any>[]>>;

    /**
     * Calculates the token budget for the prompt section.
     * @remarks
     * If the section has a fixed length, the budget will be the minimum of the section's length
     * and the maximum number of tokens. Otherwise, the budget will be the maximum number of tokens.
     * @param maxTokens Maximum number of tokens allowed for the rendered prompt.
     * @returns The token budget for the prompt section.
     */
    protected getTokenBudget(maxTokens: number): number {
        return this.tokens > 1.0 ? Math.min(this.tokens, maxTokens) : maxTokens;
    }

    /**
     * Helper method for returning a list of messages from `renderAsMessages()`.
     * @remarks
     * If the section has a fixed length, the function will truncate the list of messages to
     * fit within the token budget.
     * @param output List of messages to return.
     * @param length Total number of tokens consumed by the list of messages.
     * @param tokenizer Tokenizer to use for encoding/decoding text.
     * @param maxTokens Maximum number of tokens allowed for the rendered prompt.
     * @returns The rendered prompt section.
     */
    protected returnMessages(output: Message[], length: number, tokenizer: Tokenizer, maxTokens: number): RenderedPromptSection<Message[]> {
        // Truncate if fixed length
        if (this.tokens > 1.0) {
            while (length > this.tokens) {
                const msg = output.pop();
                const encoded = tokenizer.encode(PromptSectionBase.getMessageText(msg!));
                length -= encoded.length;
                if (length < this.tokens) {
                    const delta = this.tokens - length;
                    const truncated = tokenizer.decode(encoded.slice(0, delta));
                    output.push({ role: msg!.role, content: truncated });
                    length += delta;
                }
            }
        }

        return { output: output, length: length, tooLong: length > maxTokens };
    }

    /**
     * Returns the content of a message as a string.
     * @param message Message to get the text of.
     * @returns The message content as a string.
     */
    public static getMessageText(message: Message): string {
        let text: MessageContentParts[]|string = message.content ?? '';
        if (Array.isArray(text)) {
            text = text.filter((part) => part.type === 'text').map((part) => part.text).join(' ');
        } else if (message.function_call) {
            text = JSON.stringify(message.function_call);
        } else if (message.name) {
            text = `${message.name} returned ${text}`;
        }

        return text;
    }
}

File: python/packages/ai/teams/ai/prompts/sections/prompt_section_base.py
"""
Copyright (c) Microsoft Corporation. All rights reserved.
Licensed under the MIT License.
"""

from __future__ import annotations

import json
from abc import abstractmethod
from dataclasses import asdict
from typing import Any, List, Union

from botbuilder.core import TurnContext

from ....state import MemoryBase
from ...tokenizers import Tokenizer
from ..message import Message, MessageContentParts, TextContentPart
from ..prompt_functions import PromptFunctions
from ..rendered_prompt_section import RenderedPromptSection
from .prompt_section import PromptSection


class PromptSectionBase(PromptSection):
    """
    Abstract Base class for most prompt sections.

    This class provides a default implementation of `renderAsText()` so that derived classes only
    need to implement `renderAsMessages()`.

    Attributes:
        required (bool): If true the section is mandatory otherwise it can be safely dropped.
        tokens (float): The requested token budget for this section.
        separator (str): The separator to use between messages when rendering as text.
        text_prefix (str): The prefix to use when rendering as text.
    """

    def __init__(
        self,
        tokens: float = -1,
        required: bool = True,
        separator: str = "\n",
        text_prefix: str = "",
    ):
        """Initializes the PromptSectionBase.

        Args:
            tokens (float, optional): Sizing strategy for this section. Defaults to `auto`.
            required (bool, optional): Indicates if this section is required. Defaults to `true`.
            separator (str, optional): Separator to use between sections when rendering as text.
              Defaults to `\n`.
            text_prefix (str, optional): Prefix to use for text output. Defaults to ``.
        """
        self._required = required
        self._tokens = tokens
        self._separator = separator
        self._text_prefix = text_prefix

    @property
    def required(self) -> bool:
        """
        If true the section is mandatory otherwise it can be safely dropped.
        """
        return self._required

    @property
    def tokens(self) -> float:
        """
        The requested token budget for this section.
        - Values between 0.0 and 1.0 represent a percentage of the total budget
          and the section will be layed out proportionally to all other sections.
        - Values greater than 1.0 represent the max number of tokens the section
          should be allowed to consume.
        """
        return self._tokens

    @property
    def separator(self) -> str:
        """
        The separator to use between messages when rendering as text.
        """
        return self._separator

    @property
    def text_prefix(self) -> str:
        """
        The prefix to use when rendering as text.
        """
        return self._text_prefix

    @abstractmethod
    async def render_as_messages(
        self,
        context: TurnContext,
        memory: MemoryBase,
        functions: PromptFunctions,
        tokenizer: Tokenizer,
        max_tokens: int,
    ) -> RenderedPromptSection[List[Message]]:
        """
        Renders the section as a list of messages.

        Args:
            context (TurnContext): Context for the current turn of conversation with the user.
            memory (MemoryBase): An interface for accessing state values.
            functions (PromptFunctions): Registry of functions that can be used by the section.
            tokenizer (Tokenizer): Tokenizer to use when rendering the section.
            max_tokens (int): Maximum number of tokens allowed to be rendered.

        Returns:
            RenderedPromptSection[List[Message]]: The rendered prompt section as a list of messages.
        """

    async def render_as_text(
        self,
        context: TurnContext,
        memory: MemoryBase,
        functions: PromptFunctions,
        tokenizer: Tokenizer,
        max_tokens: int,
    ) -> RenderedPromptSection[str]:
        """
        Renders the section as a string of text.

        Args:
            context (TurnContext): Context for the current turn of conversation with the user.
            memory (MemoryBase): An interface for accessing state values.
            functions (PromptFunctions): Registry of functions that can be used by the section.
            tokenizer (Tokenizer): Tokenizer to use when rendering the section.
            max_tokens (int): Maximum number of tokens allowed to be rendered.

        Returns:
            RenderedPromptSection[str]: The rendered prompt section as a string.
        """

        # Render as messages
        as_messages: RenderedPromptSection[List[Message[Any]]] = await self.render_as_messages(
            context, memory, functions, tokenizer, max_tokens
        )
        if len(as_messages.output) == 0:
            return RenderedPromptSection("", 0, False)

        # Convert to text
        text = self.separator.join(
            [PromptSectionBase.get_message_text(message) for message in as_messages.output]
        )

        # Calculate length
        prefix_length = len(tokenizer.encode(self.text_prefix))
        separator_length = len(tokenizer.encode(self.separator))
        length = (
            prefix_length + as_messages.length + (len(as_messages.output) - 1) * separator_length
        )

        # Truncate if fixed length
        text = self.text_prefix + text
        if self.tokens > 1.0 and length > self.tokens:
            encoded = tokenizer.encode(text)
            text = tokenizer.decode(encoded[: int(self.tokens)])
            length = int(self.tokens)

        return RenderedPromptSection(text, length, length > max_tokens)

    def _get_token_budget(self, max_tokens: int) -> int:
        return min(int(self.tokens), max_tokens) if self.tokens > 1.0 else max_tokens

    def _return_messages(
        self, output: List[Message], length: int, tokenizer: Tokenizer, max_tokens: int
    ) -> RenderedPromptSection:
        # Truncate if fixed length
        if self.tokens > 1.0:
            while length > self.tokens:
                msg = output.pop()
                encoded = tokenizer.encode(PromptSectionBase.get_message_text(msg))
                length -= len(encoded)
                if length < self.tokens:
                    delta = self.tokens - length
                    truncated = tokenizer.decode(encoded[: int(delta)])
                    output.append(Message(msg.role, truncated))
                    length += int(delta)

        return RenderedPromptSection(output, length, length > max_tokens)

    @staticmethod
    def get_message_text(message: Message) -> str:
        """
        Returns the content of a message as a string.

        Args:
            message (Message): Message to get the text of.

        Returns:
            str: The message content as a string.
        """
        text: Union[List[MessageContentParts], str] = message.content or ""
        if isinstance(text, list):
            text = " ".join([part.text for part in text if isinstance(part, TextContentPart)])
        elif message.function_call:
            text = json.dumps(asdict(message.function_call))
        elif message.name:
            text = f"{message.name} returned {text}"

        return text
